api/api-doc/storage_service.json

```diff
@@ -2638,6 +2638,70 @@
             }
          ]
       },
+      {
+         "path":"/storage_service/tablets/del_replica",
+         "operations":[
+            {
+               "nickname":"del_tablet_replica",
+               "method":"POST",
+               "summary":"Deletes replica from tablet",
+               "type":"void",
+               "produces":[
+                  "application/json"
+               ],
+               "parameters":[
+                  {
+                     "name":"ks",
+                     "description":"Keyspace name",
+                     "required":true,
+                     "allowMultiple":false,
+                     "type":"string",
+                     "paramType":"query"
+                  },
+                  {
+                     "name":"table",
+                     "description":"Table name",
+                     "required":true,
+                     "allowMultiple":false,
+                     "type":"string",
+                     "paramType":"query"
+                  },
+                  {
+                     "name":"token",
+                     "description":"Token owned by the tablet to delete replica from",
+                     "required":true,
+                     "allowMultiple":false,
+                     "type":"integer",
+                     "paramType":"query"
+                  },
+                  {
+                     "name":"host",
+                     "description":"Host id to remove replica from",
+                     "required":true,
+                     "allowMultiple":false,
+                     "type":"string",
+                     "paramType":"query"
+                  },
+                  {
+                     "name":"shard",
+                     "description":"Shard number to remove replica from",
+                     "required":true,
+                     "allowMultiple":false,
+                     "type":"integer",
+                     "paramType":"query"
+                  },
+                  {
+                     "name":"force",
+                     "description":"When set to true, replication strategy constraints can be broken (false by default)",
```

Comment:
This means we can even remove the last replica of a tablet. E.g., we have RF 3, we can remove all 3 of them, resulting zero replica for a given token range. Will this break some of the assumptions in the whole code base?

Similarly, with the add replica api and RF 3, can we add 5 replicas to a given tablet?

Reply:
> E.g., we have RF 3, we can remove all 3 of them, resulting zero replica for a given token range. Will this break some of the assumptions in the whole code base?
> Similarly, with the add replica api and RF 3, can we add 5 replicas to a given tablet?

One-by-one. Yes, we can. Likely this can make CL=QUORUM to stop working the expected way, as it checks keyspace's RF, not the actual amount of replicas.

This PR intent is to facilitate larger work with RF change via CQL. The API endpoint in question is mainly for newly introduced test and is to be used with care, of course

Reply:
This api is exposed to users. It can be misused easily. We need to find a way to prevent user to shoot itself with this api somehow. 

At least, we should document it in the api description saying it might cause consistency issues. 

@bhalevy @tgrabiec

Reply:
Test APIs should not be enabled in production.
We can limit them to non-release build mode or have a configuration option to enable them so it'd be clear to the user that they are enabling unsafe APIs that provide backdoors for error injection.

Reply:
> Likely this can make CL=QUORUM to stop working the expected way, as it checks keyspace's RF, not the actual amount of replicas.

Yes, looking at `filter_for_query` -> `block_for` -> `quorum_for` I see that the number of nodes to query for quorum is (indirectly) taken from keyspace configuration, not from actual replica set

> We need to find a way to prevent user to shoot itself with this api somehow.
> We can limit them to non-release build mode or have a configuration option to enable them so it'd be clear to the user that they are enabling unsafe APIs that provide backdoors for error injection.

User (given enough rights) can directly modify system.tablets breaking things up the way he wants, isn't it?
I'm for documenting that the API is dangerous and is to be used only if 101% aware of the consequences.

Reply:
> One-by-one. Yes, we can. Likely this can make CL=QUORUM to stop working the expected way, as it checks keyspace's RF, not the actual amount of replicas.

Maybe that's the expected way actually. Removing a replica is like having it permanently down. You have to add back replicas for CL to be met again.